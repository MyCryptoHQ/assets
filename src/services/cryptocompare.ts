import { Asset } from '../constants';
import { CachedResponse, getInitialCache, withCache } from './cache';
import { isSimilar } from '../utils';

const CRYPTOCOMPARE_API_ENDPOINT = 'https://min-api.cryptocompare.com/data/all/coinlist';

interface CryptoCompareData {
  Data: {
    [key: string]: {
      Id: string;
      CoinName: string;
      SmartContractAddress: string;
      BuiltOn: string;
      Symbol: string;
    };
  };
}

/**
 * Match an asset with a CryptoCompare ID.
 *
 * @param {Partial<CryptoCompareData>} cache
 * @param {Asset} asset
 * @return {Promise<CachedResponse<CryptoCompareData, string | null>>}
 */
export const matchCryptoCompareId = async (
  cache: Partial<CryptoCompareData>,
  asset: Asset
): Promise<CachedResponse<CryptoCompareData, string | null>> => {
  let data = cache;
  if (!data || !data.Data) {
    data = await getInitialCache<CryptoCompareData>(CRYPTOCOMPARE_API_ENDPOINT);
  }

  // Search based on contract address
  if (asset.address) {
    const item = Object.values(data.Data!).find(
      token =>
        (token.SmartContractAddress && token.SmartContractAddress.toLowerCase()) ===
        asset.address!.toLowerCase()
    );

    if (item && item.BuiltOn === '7605') {
      return {
        cache: data,
        data: item.Symbol
      };
    }
  }

  // Search based on name and symbol
  if (data.Data![asset.symbol] && isSimilar(data.Data![asset.symbol].CoinName, asset.name)) {
    return {
      cache: data,
      data: data.Data![asset.symbol].Symbol
    };
  }

  return {
    cache: data,
    data: null
  };
};

export default withCache<CryptoCompareData, typeof matchCryptoCompareId, string | null>(
  matchCryptoCompareId
);
