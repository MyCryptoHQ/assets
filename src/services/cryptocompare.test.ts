import { matchCryptoCompareId } from './cryptocompare';
import { getInitialCache } from './cache';

const CACHE_DATA = {
  ETH: {
    Id: '7605',
    Name: 'ETH',
    Symbol: 'ETH',
    CoinName: 'Ethereum'
  },
  GNT: {
    Id: '33022',
    Name: 'GNT',
    Symbol: 'GNT',
    CoinName: 'Golem Network Token',
    BuiltOn: '7605',
    SmartContractAddress: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d'
  },
  FOO: {
    Id: '1337',
    Name: 'Foo',
    Symbol: 'FOO',
    CoinName: 'Foo Token',
    BuiltOn: '1',
    SmartContractAddress: '0x0000000000000000000000000000000000000000'
  }
};

jest.mock('./cache', () => ({
  getInitialCache: jest.fn(async () => ({ Data: CACHE_DATA })),

  withCache: jest.requireActual('./cache').withCache
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('matchCryptoCompareId', () => {
  it('initializes the cache', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    const { cache } = await matchCryptoCompareId({}, asset);
    const result = await matchCryptoCompareId(cache, asset);

    expect(result.cache.Data).toMatchSnapshot();
    expect(getInitialCache).toHaveBeenCalledTimes(1);
  });

  it('fetches the ID of an asset', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchCryptoCompareId({}, asset)).resolves.toEqual({
      cache: {
        Data: CACHE_DATA
      },
      data: 'ETH'
    });
  });

  it('fetches the ID of a token by address for ERC-20 tokens', async () => {
    const asset = {
      address: '0xa74476443119A942dE498590Fe1f2454d7D4aC0d',
      decimal: '18',
      name: 'Golem',
      symbol: 'GNT',
      uuid: '59a6f871-34bb-5914-86cc-b5f1307bb3b9'
    };

    await expect(matchCryptoCompareId({}, asset)).resolves.toEqual({
      cache: {
        Data: CACHE_DATA
      },
      data: 'GNT'
    });
  });

  it('does not fetch the ID of a token for non-ERC-20 tokens', async () => {
    const asset = {
      address: '0x0000000000000000000000000000000000000000',
      decimal: '18',
      name: 'Bar Token',
      symbol: 'BAR',
      uuid: '5b756932-ed7e-47c1-a8b8-72fb5ec7e229'
    };

    await expect(matchCryptoCompareId({}, asset)).resolves.toEqual({
      cache: {
        Data: CACHE_DATA
      },
      data: null
    });
  });

  it('returns null if an asset could not be found', async () => {
    const asset = {
      name: 'Ethereum Classic',
      symbol: 'ETC',
      uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
    };

    await expect(matchCryptoCompareId({}, asset)).resolves.toEqual({
      cache: {
        Data: CACHE_DATA
      },
      data: null
    });
  });
});
