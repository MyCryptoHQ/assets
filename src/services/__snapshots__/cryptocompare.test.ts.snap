// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchCryptoCompareId initializes the cache 1`] = `
Object {
  "ETH": Object {
    "CoinName": "Ethereum",
    "Id": "7605",
    "Name": "ETH",
    "Symbol": "ETH",
  },
  "FOO": Object {
    "BuiltOn": "1",
    "CoinName": "Foo Token",
    "Id": "1337",
    "Name": "Foo",
    "SmartContractAddress": "0x0000000000000000000000000000000000000000",
    "Symbol": "FOO",
  },
  "GNT": Object {
    "BuiltOn": "7605",
    "CoinName": "Golem Network Token",
    "Id": "33022",
    "Name": "GNT",
    "SmartContractAddress": "0xa74476443119A942dE498590Fe1f2454d7D4aC0d",
    "Symbol": "GNT",
  },
}
`;
